src/object_store/api.nim

```diff
@@ -0,0 +1,148 @@
+
+proc canonicalizeKey(key: string, data: Box): Box =
+  let callbackOpt = attrGetOpt[CallbackObj]("keyspec." & key & ".canonicalize")
+  if callbackOpt.isNone():
+    trace("object store: no canonicalize() for " & key)
+    return data
+  let callback = callbackOpt.get()
+  trace("object store: canonicalizing " & key & " with " & $callback)
+  try:
+    let canonicalized = runCallback(callback, @[data])
+    #let canonicalized = runCallback("keyspec." & key & ".canonicalize", @[data])
+    if canonicalized.isNone():
+      error("object store: missing implementation to canonicalize " & key & " for " & $callback)
+      return data
+    return canonicalized.get()
+  except:
+    error("object store: failed to canonicalize " & key & " due to:" & getCurrentExceptionMsg())
+    dumpExOnDebug()
+    return data
+
+proc objectifyKey(dict:            ChalkDict,
+                  k:               string,
+                  v:               Box,
+                  content:         string,
+                  objectsData:     ObjectsDict,
+                  storeConfigName: string) =
+  let objectRefs = objectsData.mgetOrPut(storeConfigName, newOrderedTable[string, ObjectStoreRef]())
+  # key was already objectified before. use existing ref
+  if k in objectRefs:
+    dict[objectStorePrefix & k] = pack($(objectRefs[k]))
+    return
+  let
+    storeConfig = getObjectStoreConfigByName(storeConfigName)
+    canonical   = canonicalizeKey(k, v)
+    lookupRef   = newObjectStoreRef(storeConfig, k, content, canonical)
+  var storeRef: ObjectStoreRef
+  try:
+    trace("object store: looking up key in object store: " & $lookupRef)
+    storeRef = storeConfig.store.objectExists(storeConfig, lookupRef)
+    if storeRef == nil:
+      try:
+        trace("object store: creating key in object store: " & $lookupRef)
+        storeRef = storeConfig.store.createObject(storeConfig, lookupRef, content)
+      except:
+        let e = getCurrentExceptionMsg()
+        # it is possible multiple chalks might be trying to upload the same object/hash
+        # to the object store at the same time in which case only one should succeed
+        # and so just in case attempt to fetch existing object again
+        # p.s. this is more likely for externally created metadata
+        # where there could be many builds referencing the same metadata
+        # such as _IMAGE_SBOM for a base image like alpine
+        trace("object store: attempting to refetch object to handle possible creation race conditions")
+        storeRef = storeConfig.store.objectExists(storeConfig, lookupRef)
+        if storeRef == nil:
+          error("object store: could not upload object " & $lookupRef & " due to: " & e)
+          dumpExOnDebug()
+  except:
+    error("object store: could not lookup existing object " & $lookupRef & " due to: " & getCurrentExceptionMsg())
+    dumpExOnDebug()
+  if storeRef != nil:
+    objectRefs[k] = storeRef
+    dict[objectStorePrefix & k] = pack($storeRef)
```

Comment:
should this be the `else` to balance the `if` on line 98?

Reply:
no

this is reference the var from:

https://github.com/crashappsec/chalk/blob/962a9b6698f2107234058746b6284a9027faec88/src/object_store/api.nim#L94

then inside the try statement various places attempt to set it if possible:

* query existing object
* if fails, attempt to upload object
* if fails attempt to query existing object again to handle concurrent upload race conditions

then here if any of them succeeded, add the reference to the `dict`. its not related to the if statements which define the fallback procedures for setting the `storeRef` in general

Reply:
understood. thx!